flow/hooks/track_operations.py

```diff
@@ -0,0 +1,164 @@
+
+        def _log_operation(operation, job, error=None):
+            if self.strict_git:
+                if git.Repo(job._project.path).is_dirty():
+                    raise RuntimeError(
+                        "Unable to reliably log operation, because the git repository in "
+                        "the project root directory is dirty.\n\nMake sure to commit all "
+                        "changes or ignore this warning by setting '{}(strict_git=False)'.".format(
+                            type(self).__name__
+                        )
+                    )
+            metadata = collect_metadata(operation, job)
+
+            # Add additional execution-related information to metadata.
+            metadata["stage"] = stage
+            metadata["error"] = None if error is None else str(error)
+
+            # Write metadata to collection inside job workspace.
+            with open(job.fn(FN_LOGFILE), "a") as logfile:
+                # HELP: Is there any reason we used Collections instead of a regular dictionary?
+                logfile.write(json.dumps(metadata) + "\n")
+
+        return _log_operation
+
+    def on_start(self, operation, job):
+        """Track the start of execution of an operation on a job."""
+        self.log_operation(stage="prior")(operation, job)
+
+    def on_success(self, operation, job):
+        """Track the successful completion of an operation on a job."""
+        self.log_operation(stage="after")(operation, job)
+
+    def on_exception(self, operation, error, job):
+        """Log errors raised during the execution of an operation on a job."""
+        self.log_operation(stage="after")(operation, job, error)
+
+    def install_operation_hooks(self, op, project_cls=None):
+        """Decorate operation to track execution.
+
+        Parameters
+        ----------
+        op : function or type
+            An operation function to log or a subclass of `flow.FlowProject` if ``project_cls`` is
+            ``None``.
+        project_cls : type
+            A subclass of `flow.FlowProject`.
+        """
+        if project_cls is None:
+            return lambda func: self.install_operation_hooks(func, op)
+        project_cls.operation_hooks.on_start(self.on_start)(op)
+        project_cls.operation_hooks.on_success(self.on_success)(op)
+        project_cls.operation_hooks.on_exception(self.on_exception)(op)
+        return op
+
+    def install_project_hooks(self, project):
+        """Install track operation to all operations in a Signac-flow project.
+
+        Parameters
+        ----------
+        project : flow.FlowProject
+            The project to install hooks on.
+        """
+        project.project_hooks.on_start.append(self.on_start)
+        project.project_hooks.on_success.append(self.on_success)
+        project.project_hooks.on_exception.append(self.on_exception)
+        return project
+
+    __call__ = install_project_hooks
```

Comment:
I’m not sure how to feel about having a call operator AND a method that do the same thing. Do we need both? Why?